Description: DOS-style DSZLOG support
 It is useful when using lrzsz with BBS software packages.
Author: Ryan Underwood <nemesis@dbz.icequake.net>
Bug-Debian: https://bugs.debian.org/206648

--- lrzsz-0.12.21.orig/src/lrz.c
+++ lrzsz-0.12.21/src/lrz.c
@@ -69,6 +69,7 @@ extern int errno;
 unsigned Baudrate = 2400;
 
 FILE *fout;
+char *dszlogpath;
 
 
 int Lastrx;
@@ -193,6 +194,9 @@ bibi(int n)
 		zmputs(Attn);
 	canit(STDOUT_FILENO);
 	io_mode(0,0);
+	if (dszlogpath) {
+		dszlog(dszlogpath, 'L',	0, 0, 0, errors, 0, 0);
+	}
 	error(128+n,0,_("caught signal %d; exiting"), n);
 }
 
@@ -266,6 +270,13 @@ main(int argc, char *argv[])
 	if ((cp=getenv("ZMODEM_RESTRICTED"))!=NULL)
 		Restricted=2;
 
+	dszlogpath = getenv("DSZLOG");
+	if (access(dszlogpath, F_OK) == 0) {
+		if (unlink(dszlogpath) != 0) {
+			perror("unlink dszlog");
+		}
+	}
+
 	/* make temporary and unfinished files */
 	umask(0077);
 
@@ -560,6 +571,7 @@ main(int argc, char *argv[])
 		signal(SIGINT, bibi);
 	signal(SIGTERM, bibi);
 	signal(SIGPIPE, bibi);
+	signal(SIGHUP, bibi);
 	if (wcreceive(npats, patts)==ERROR) {
 		exitcode=0200;
 		canit(STDOUT_FILENO);
@@ -686,9 +698,9 @@ wcreceive(int argc, char **argp)
 #endif
 			if (c)
 				goto fubar;
-		} else {
+		} else {	/* Not Xmodem, not Zmodem, so Ymodem */
 			for (;;) {
-				if (Verbose > 1
+				if (Verbose > 1 || dszlogpath
 #ifdef ENABLE_SYSLOG
 					|| enable_syslog
 #endif
@@ -713,7 +725,7 @@ wcreceive(int argc, char **argp)
 				if (wcrx(&zi)==ERROR)
 					goto fubar;
 
-				if (Verbose > 1
+				if (Verbose > 1 || dszlogpath
 #ifdef ENABLE_SYSLOG
 					|| enable_syslog
 #endif
@@ -735,16 +747,19 @@ wcreceive(int argc, char **argp)
 						lsyslog(LOG_INFO,"%s/%s: %ld Bytes, %ld BPS",
 							shortname,protname(),zi.bytes_received, bps);
 #endif
+					if (dszlogpath)	dszlog(dszlogpath, 'R',
+								zi.bytes_received, zi.fname,
+								bps, errors, 0, 0);
 				}
 			}
 		}
-	} else {
+	} else {	/* Xmodem */
 		char dummy[128];
 		dummy[0]='\0'; /* pre-ANSI HPUX cc demands this */
 		dummy[1]='\0'; /* procheader uses name + 1 + strlen(name) */
 		zi.bytes_total = DEFBYTL;
 
-		if (Verbose > 1
+		if (Verbose > 1 || dszlogpath
 #ifdef ENABLE_SYSLOG
 			|| enable_syslog
 #endif
@@ -783,7 +798,7 @@ wcreceive(int argc, char **argp)
 		if (wcrx(&zi)==ERROR) {
 			goto fubar;
 		}
-		if (Verbose > 1
+		if (Verbose > 1 || dszlogpath
 #ifdef ENABLE_SYSLOG
 			|| enable_syslog
 #endif
@@ -804,6 +819,10 @@ wcreceive(int argc, char **argp)
 				lsyslog(LOG_INFO,"%s/%s: %ld Bytes, %ld BPS",
 					shortname,protname(),zi.bytes_received, bps);
 #endif
+			if (dszlogpath)	dszlog(dszlogpath, 'R',
+						zi.bytes_received, zi.fname,
+						bps, errors, 0, 0);
+
 		}
 	}
 	return OK;
@@ -813,6 +832,13 @@ fubar:
 		lsyslog(LOG_ERR,"%s/%s: got error", 
 			shortname ? shortname : "no.name", protname());
 #endif
+	if (dszlogpath)	{
+		double d = timing(0,NULL);
+		dszlog(dszlogpath, 'E', zi.bytes_received, zi.fname,
+			(zi.bytes_received-zi.bytes_skipped)/((d) ? d : 0.5),
+			errors, 0, 0);
+	}
+
 	canit(STDOUT_FILENO);
 	if (Topipe && fout) {
 		pclose(fout);  return ERROR;
@@ -1784,10 +1810,18 @@ static int
 rzfiles(struct zm_fileinfo *zi)
 {
 	register int c;
+	long bps;
 
 	for (;;) {
 		timing(1,NULL);
 		c = rzfile(zi);
+		{
+			double d;
+			d=timing(0,NULL);
+			if (d==0)
+				d=0.5; /* can happen if timing uses time() */
+			bps=(zi->bytes_received-zi->bytes_skipped)/d;
+		}
 		switch (c) {
 		case ZEOF:
 			if (Verbose > 1
@@ -1795,12 +1829,6 @@ rzfiles(struct zm_fileinfo *zi)
 				|| enable_syslog
 #endif
 	 		) {
-				double d;
-				long bps;
-				d=timing(0,NULL);
-				if (d==0)
-					d=0.5; /* can happen if timing uses time() */
-				bps=(zi->bytes_received-zi->bytes_skipped)/d;
 				if (Verbose > 1) {
 					vstringf(
 						_("\rBytes received: %7ld/%7ld   BPS:%-6ld                \r\n"),
@@ -1808,6 +1836,8 @@ rzfiles(struct zm_fileinfo *zi)
 				}
 				DO_SYSLOG_FNAME((LOG_INFO, "%s/%s: %ld Bytes, %ld BPS",shortname,
 						   protname(), (long) zi->bytes_total,bps));
+				if (dszlogpath) dszlog(dszlogpath, 'z', zi->bytes_received,
+						zi->fname, bps, errors, 0, 0);
 			}
 			/* FALL THROUGH */
 		case ZSKIP:
@@ -1816,6 +1846,8 @@ rzfiles(struct zm_fileinfo *zi)
 				if (Verbose) 
 					vstringf(_("Skipped"));
 				DO_SYSLOG_FNAME((LOG_INFO, "%s/%s: skipped",shortname,protname()));
+				if (dszlogpath) dszlog(dszlogpath, 'E', zi->bytes_received,
+						zi->fname, bps, errors, 0, 0);
 			}
 			switch (tryz()) {
 			case ZCOMPL:
@@ -1830,6 +1862,8 @@ rzfiles(struct zm_fileinfo *zi)
 			return c;
 		case ERROR:
 			DO_SYSLOG_FNAME((LOG_INFO, "%s/%s: error",shortname,protname()));
+			if (dszlogpath) dszlog(dszlogpath, 'E', zi->bytes_received,
+					zi->fname, bps, errors, 0, 0);
 			return ERROR;
 		}
 	}
--- lrzsz-0.12.21.orig/src/lsyslog.c
+++ lrzsz-0.12.21/src/lsyslog.c
@@ -28,6 +28,8 @@
 #include <string.h>
 #endif
 
+#include <stdio.h>
+
 #if __STDC__
 #  include <stdarg.h>
 #  define VA_START(args, lastarg) va_start(args, lastarg)
@@ -81,3 +83,23 @@ lsyslog(prio,format,va_alist)
 #endif
 }
 
+void dszlog(char *logname, char status, unsigned long num_bytes, char *fname,
+		int cps, int retry_events, int flow_events, int blocksize) {
+
+	FILE *mylog;
+	if ((mylog = fopen(logname, "a"))) {
+		fprintf(mylog, "%c %6ld %5ld bps %4ld cps %3d errors %5u %4d %s %ld\r\n",
+				status,
+				num_bytes,
+				cps*8, /* XXX where to get DTE from? */
+				cps,
+				retry_events,
+				flow_events, /* XXX where is flow_control? */
+				blocksize,
+				fname,
+				-1 /* serial number */
+				);
+		fclose(mylog);
+	}
+}
+
--- lrzsz-0.12.21.orig/src/lsz.c
+++ lrzsz-0.12.21/src/lsz.c
@@ -60,6 +60,8 @@ extern char *strerror();
 extern int errno;
 #endif
 
+char *dszlogpath;
+
 unsigned Baudrate=2400;	/* Default, should be set by first mode() call */
 unsigned Txwindow;	/* Control the size of the transmitted window */
 unsigned Txwspac;	/* Spacing between zcrcq requests */
@@ -224,6 +226,9 @@ bibi (int n)
 	canit(STDOUT_FILENO);
 	fflush (stdout);
 	io_mode (io_mode_fd,0);
+	if (dszlogpath) {
+		dszlog(dszlogpath, 'L', 0, 0, 0, errors, 0, 0);
+	}
 	if (n == 99)
 		error (0, 0, _ ("io_mode(,2) in rbsb.c not implemented\n"));
 	else
@@ -333,6 +338,14 @@ main(int argc, char **argv)
 	}
 	if ((cp=getenv("ZMODEM_RESTRICTED"))!=NULL)
 		Restricted=1;
+
+	dszlogpath = getenv("DSZLOG");
+	if (access(dszlogpath, F_OK) == 0) {
+		if (unlink(dszlogpath) != 0) {
+			perror("unlink dszlog");
+		}
+	}
+
 	from_cu();
 	chkinvok(argv[0]);
 
@@ -1114,6 +1127,12 @@ wcs(const char *oname, const char *remot
 		if (enable_syslog)
 			lsyslog(LOG_INFO, _("%s/%s: error occured"),protname(),shortname);
 #endif
+		if (dszlogpath) {
+			double d = timing(0,NULL);
+			dszlog(dszlogpath, 'E', zi.bytes_sent, zi.fname,
+				zi.bytes_sent/((d) ? d : 0.5),
+				errors, 0, blklen);
+		}
 		return ERROR;
 	case ZSKIP:
 		error(0,0, _("skipped: %s"),name);
@@ -1121,6 +1140,13 @@ wcs(const char *oname, const char *remot
 		if (enable_syslog)
 			lsyslog(LOG_INFO, _("%s/%s: skipped"),protname(),shortname);
 #endif
+		if (dszlogpath) {
+			double d = timing(0,NULL);
+			/* XXX is Zmodem skip really deserving a 'E' ? */
+			dszlog(dszlogpath, 'E', zi.bytes_sent, zi.fname,
+				zi.bytes_sent/((d) ? d : 0.5),
+				errors, 0, blklen);
+		}
 		return OK;
 	}
 	if (!zmodem_requested && wctx(&zi)==ERROR)
@@ -1129,12 +1155,18 @@ wcs(const char *oname, const char *remot
 		if (enable_syslog)
 			lsyslog(LOG_INFO, _("%s/%s: error occured"),protname(),shortname);
 #endif
+		if (dszlogpath) {
+			double d = timing(0,NULL);
+			dszlog(dszlogpath, 'E', zi.bytes_sent, zi.fname,
+				zi.bytes_sent/((d) ? d : 0.5),
+				errors, 0, blklen);
+		}
 		return ERROR;
 	}
 	if (Unlinkafter)
 		unlink(oname);
 
-	if (Verbose > 1
+	if (Verbose > 1 || dszlogpath
 #ifdef ENABLE_SYSLOG
 		|| enable_syslog
 #endif
@@ -1153,6 +1185,11 @@ wcs(const char *oname, const char *remot
 			lsyslog(LOG_INFO, "%s/%s: %ld Bytes, %ld BPS",shortname,
 				protname(), (long) zi.bytes_sent,bps);
 #endif
+		if (dszlogpath) {
+			char whichprot = (protocol==ZM_ZMODEM)? 'Z' : 'S';
+			dszlog(dszlogpath, whichprot, zi.bytes_sent,
+					zi.fname, bps, errors, 0, blklen);
+		}
 	}
 	return 0;
 }
--- lrzsz-0.12.21.orig/src/zglobal.h
+++ lrzsz-0.12.21/src/zglobal.h
@@ -456,6 +456,8 @@ int tcp_accept __P ((int d)) LRZSZ_ATTRI
 
 const char * protname __P ((void)) LRZSZ_ATTRIB_SECTION(lrzsz_rare);
 void lsyslog __P ((int, const char *,...));
+void dszlog(char*, char, unsigned long, char*, int, int, int, int);
+
 
 
 
